Boost: Cache Preloading Scheduling

LiamSarsfield opened the pull request:
> [!NOTE]
> Merge this PR after https://github.com/Automattic/jetpack/pull/42549 is merged.

<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->
<!-- a12s: If you have an expected version that you're aiming for the PR to add, please use the Milestone field to communicate it. If you leave it blank, that indicates there isn't a preference. -->

Resolves [HOG-4](https://linear.app/a8c/issue/HOG-4)

## Proposed changes:
* If Page Caching is enabled, the Cornerstone Pages are preloaded every 30 mins.
* If users upgrade to the version of Boost with this PR, and have Page Caching enabled, the schedule is created.

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->
pc9hqz-3nV-p2

## Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->
N/A

## Testing instructions:

### Prerequisites
- Ensure that the constant JETPACK_BOOST_ALPHA_FEATURES is defined as true either in an mu-plugin or wp-config.php.
- Install WP Crontrol Plugin
- If testing on a Jurassic Ninja site, ensure Edge Caching is disabled in Settings -> Edge Cache -> Disable Edge Cache
- Bypassing Edge Cache is out of scope for this PR.

### Migration Testing
1. Set up Boost on a site without this PR/branch.
2. Enable Page Caching.
3. Now install this branch. Ensure that the plugin's version has changed during this installation. This can be done either automatically through the Jetpack Beta plugin, or by changing the `JETPACK_BOOST_VERSION` constant.
4. Within WP Crontrol, verify that a new Cron has been scheduled called `jetpack_boost_preload_cornerstone`, with the interval "Twice Hourly".

### Cron Testing
1. Go to Boost Settings page.
2. Disable Page Caching
4. Within WP Crontrol, verify that the cron schedule `jetpack_boost_preload_cornerstone`, is no longer present.
5. Enable Page Cache on the WordPress site.
6. Within WP Crontrol, verify that a new Cron has been scheduled called `jetpack_boost_preload_cornerstone`, with the interval "Twice Hourly".
7. Within the WordPress site, head to ./wp-content/boost-cache/cache/(domain)
8. Verify that page caching files are present.
9. Note down when a Cornerstone Page cache file was created via stat -c '%w %n' *.
10. Now within the wp-admin WP Crontrol UI, click Run now for the `jetpack_boost_preload_cornerstone` cronjob.
11. Verify using stat -c '%w %n' * that the file creation date is now different. If the file ends in .rebuild.html, wait a moment, and it should change back to .html once the cache has been built from the preload.

### Uninstalling Cleanup
1. Because the Jetpack Beta plugin affects uninstallation hooks for the plugin, it should be deactivated, and the Jetpack Boost plugin that was installed from it should be uninstalled.
2. Within this PR, go to the Build / Create artifact for Jetpack Beta plugin, e.g. [like this](https://github.com/Automattic/jetpack/actions/runs/13975814603/job/39129403948?pr=42611)
3. Select Summary. If next to "Plugins" you see expired, re-run the workflow.
4. Select "Plugins" to download the Boost version.
5. Decompress it
6. Rename it from jetpack-boost-dev to jetpack-boost.
7. Compress it.
8. Manually install this version on your site.
9. Verify within WP Crontrol that the `jetpack_boost_preload_cornerstone` is still present.
10. Now uninstall the Boost plugin you just installed.
11. Verify within WP Crontrol that the `jetpack_boost_preload_cornerstone` is no longer present.

LiamSarsfield committed:
Page Cache: Add support for submodules with Cache Preload

Co-authored-by: Peter Petrov <dilirity@users.noreply.github.com>

LiamSarsfield committed:
changelog

LiamSarsfield committed:
Page Cache: Enhance Cache Preload with dynamic post scheduling and update mechanism

- Refactor preload mechanism to support dynamic post scheduling
- Add method to handle post updates and schedule preloading for cornerstone pages
- Implement more flexible preload post tracking using WordPress options
- Improve error handling during page preloading

LiamSarsfield committed:
Boost: Enhance Cache Preloading and Cornerstone Page Detection

- Add new method `is_cornerstone_page_by_url()` in Cornerstone_Utils to check cornerstone pages by URL
- Modify `is_cornerstone_page()` to use the new URL-based method
- Update Cache_Preload to handle cache invalidation and preloading of cornerstone pages
- Add action hook for successful cache invalidation
- Improve method names and add comments for clarity

LiamSarsfield committed:
Boost: Improve Cache Preload Handling for Cornerstone Pages

- Enhance `handle_cache_invalidation()` method to support different cache invalidation types
- Add support for full cache deletion by preloading all Cornerstone Pages
- Improve documentation with detailed comments and @since tags
- Update method signature to include invalidation type parameter
- Refactor cache preloading logic for more robust handling of cache events

LiamSarsfield committed:
Boost: Improve Cache Preload Performance and Error Handling

- Refactor page preloading to process URLs in batches of 10
- Add batch processing to reduce server load and improve reliability
- Enhance error handling with detailed logging for preload failures
- Improve HTTP response validation during page preloading
- Add methods to manage batch processing and track remaining URLs

LiamSarsfield committed:
Boost: Add Comprehensive Unit Tests for Cache Preload Module

- Create new test suite for Cache_Preload class in Jetpack Boost
- Implement comprehensive test coverage for cache preloading methods
- Test scenarios including cornerstone page preloading, cache invalidation, and scheduling
- Verify interactions with WordPress core functions and utility classes
- Ensure robust testing of preload scheduling, post update handling, and cache management

LiamSarsfield committed:
Boost: Update Cache Preload Test Suite with Direct Method Calls

- Refactor test cases to use direct method calls instead of partial mocks
- Replace Mockery-based mocking with explicit WordPress function expectations
- Improve test coverage by directly testing Cache_Preload class methods
- Simplify test setup and reduce complexity of test assertions
- Ensure consistent testing approach for cache preloading functionality

LiamSarsfield committed:
Boost: Add Cache-Busting Headers to Page Preload Requests

- Enhance page preloading by adding cache-control headers

LiamSarsfield committed:
Boost: Refactor Cache Preload Module with Feature Flag and Improvements

- Add feature flag support for cache preloading via JETPACK_BOOST_ALPHA_FEATURES constant
- Improve type hinting and method signatures in Cache_Preload class
- Update test suite to support new feature flag implementation

LiamSarsfield committed:
Boost: Implement Always-On Cache Preload and Improve Scheduling Logic

- Add `Is_Always_On` interface to Cache_Preload class
- Enhance cron job scheduling comment to explain 2-second delay rationale

LiamSarsfield committed:
Boost: Add Cache Preload Queue Management Class

- Implement `Cache_Preload_Queue_Manager` to handle preload queue operations
- Add methods for managing, adding, and processing preload URL batches
- Provide queue management functionality with clear, modular design

LiamSarsfield committed:
Boost: Optimize Cache Preload Queue Management

- Add in-memory caching for preload queue to reduce database calls
- Improve `get_posts_to_preload()` and `set_posts_to_preload()` methods
- Refactor `add_to_queue()` to use the new in-memory cache mechanism
- Simplify queue management with a single source of truth

LiamSarsfield committed:
Boost: Refactor Cache Preload Queue Manager with Batch Size Configuration

- Add constructor to set configurable batch size for preloading
- Update `prepare_next_batch()` method to use instance batch size
- Remove redundant batch size parameter from method signature

LiamSarsfield committed:
Enhance Cache Preload Documentation and Logic

- Update comments in Cache Preload class to clarify cache management processes
- Refactor conditional checks in File Storage class for improved readability and logic flow
- Ensure accurate descriptions of cache invalidation and rebuilding processes

LiamSarsfield committed:
Remove Cache Preload Queue Manager and Refactor Cache Preload Logic

- Delete Cache_Preload_Queue_Manager class to simplify preload management
- Refactor Cache_Preload class to directly handle preload queue operations
- Clear preload queue after processing to ensure accurate state management

LiamSarsfield committed:
Update Cache Preload Scheduling Logic to Remove Delay

- Adjusted the scheduling of the preload cron job to occur immediately instead of with a 2-second delay

LiamSarsfield committed:
Refactor Cache Preload Logic to Use Direct Page Parameter

- Removed unnecessary variable assignment for the page URL in the preload_page method

dilirity committed:
Fix My jetpack assets enqueued on every page instead of just Boost's

dilirity committed:
add changelog

LiamSarsfield committed:
Merge remote-tracking branch 'origin/trunk' into add/boost/cache-preloading-cornerstone-pages

LiamSarsfield committed:
Refactor Cache Preload Methods to Accept Post URLs as Parameters

- Modified preload_pages to accept an array of posts as a parameter.
- Removed the old schedule_preload method and its associated logic.
- Adjusted handle_post_update to directly schedule preload for cornerstone pages.

LiamSarsfield committed:
Update Cache Preload to Use Array Parameters for Scheduling

- Refactored schedule_preload_cronjob to accept an array of posts.
- Adjusted tests to reflect changes in method signatures and ensure proper scheduling behavior.

LiamSarsfield committed:
Update test method names in Cache_Preload_Test to reflect cronjob functionality

- Renamed test methods to clarify that they are testing the schedule_preload_cronjob functionality.

LiamSarsfield committed:
Update Cache_Preload class to modify cronjob scheduling logic and streamline documentation

- Changed the return value of Cache Preloading feature check to false if not having features flag.
- Simplified comments for scheduling and preloading methods by removing unnecessary details.

LiamSarsfield committed:
Enhance Cache_Preload and Page_Cache classes with singleton pattern and activation logic

- Introduced a singleton pattern in Cache_Preload class with a private static instance and a public get_instance method.
- Updated Page_Cache class to implement Has_Activate interface and added an activate method to schedule cornerstone preload upon activation.

LiamSarsfield committed:
Add changelog entry for boost cache preloading feature activation

LiamSarsfield committed:
Merge remote-tracking branch 'origin/trunk' into add/boost/cache-preloading-on-cache-activation

LiamSarsfield committed:
Enhance Cache Preload Activation Logic

- Replace direct call to `Cache_Preload::get_instance()->schedule_cornerstone_preload()` with action hook `jetpack_boost_page_cache_activate` in the `activate` method of `Page_Cache` class.
- Remove singleton instance management from `Cache_Preload` class to simplify the preload logic and improve maintainability.
- Update the `setup` method in `Cache_Preload` to include the new action hook for scheduling cornerstone preloads on activation.

LiamSarsfield committed:
Refactor Cache Preload Logic for Cornerstone Pages

- Introduce new methods for scheduling and rebuilding cache for Cornerstone Pages, enhancing the cache management process.
- Update action hooks in the `setup` method to reflect changes in scheduling logic.
- Improve documentation and comments for clarity on cache rebuilding and preloading mechanisms.
- Ensure that cache invalidation and preloading are handled more efficiently through the new methods.

LiamSarsfield committed:
Add Cache Preload Activation Method and Update Class Implementations

- Introduced `get_or_create_instance_of` method in the `Setup` class to manage class instances more effectively.
- Updated `Cache_Preload` class to implement `Has_Activate` interface and added an `activate` method that utilizes the new instance management method for scheduling cornerstone preloads.
- Removed redundant `activate` method from `Page_Cache` class to streamline activation logic and ensure proper integration with the new preload scheduling mechanism.

LiamSarsfield committed:
Update Cache Preload Activation Documentation

- Added a comment to the `activate` method in the `Cache_Preload` class to clarify that activation is triggered by the parent module

LiamSarsfield committed:
Refactor `get_or_create_instance_of` Method in Setup Class

- Updated the method signature to enforce type hinting with `Has_Setup` for better clarity and type safety.
- Improved documentation to specify the expected class type for the parameter.

LiamSarsfield committed:
Remove redundant action hook for cornerstone preload scheduling in Cache_Preload setup method to streamline the activation process.

LiamSarsfield committed:
Enhance Cache Preload Functionality and Class Structure

- Introduce `get_or_create_instance_of` method in the `Setup` class to manage class instances more effectively.
- Update `Cache_Preload` class to implement `Has_Activate` interface, allowing activation logic to trigger cornerstone preload scheduling.
- Remove direct activation method from `Page_Cache` class, streamlining the activation process.
- Improve documentation and comments for better clarity on the new methods and their purposes.

LiamSarsfield committed:
Update Cache Preload Activation to Schedule Cornerstone Rebuild Cronjob

- Modify the `activate` method in the `Cache_Preload` class to call `schedule_cornerstone_rebuild_cronjob()` instead of `schedule_cornerstone_preload()`.

LiamSarsfield committed:
Merge branch 'add/boost/cache-preloading-on-cache-activation' into add/boost/cache-preloading-force-rebuild

LiamSarsfield committed:
Update `get_or_create_instance_of` Method Signature in Setup Class

- Changed the parameter type hint from `Has_Setup` to `class-string<T>` for improved type safety and clarity in documentation.

LiamSarsfield committed:
Merge branch 'add/boost/cache-preloading-on-cache-activation' into add/boost/cache-preloading-force-rebuild

LiamSarsfield committed:
Merge remote-tracking branch 'origin/trunk' into add/boost/cache-preloading-on-cache-activation

LiamSarsfield committed:
Merge branch 'add/boost/cache-preloading-on-cache-activation' into add/boost/cache-preloading-force-rebuild

LiamSarsfield committed:
Add comment to suppress Phan type mismatch warning in Setup class

- Added a comment to suppress the Phan type mismatch warning for the `add` method call in the `get_or_create_instance_of` method, ensuring clarity in type expectations.

LiamSarsfield committed:
Merge branch 'add/boost/cache-preloading-on-cache-activation' into add/boost/cache-preloading-force-rebuild

LiamSarsfield committed:
Remove unused Has_Setup interface from Cache_Preload class to simplify implementation.

LiamSarsfield committed:
Refactor Cache Preload Activation in Cache_Preload Class

- Removed the use of `Setup::get_or_create_instance_of` in the `activate` method and replaced it with direct instantiation of the `Cache_Preload` class to simplify the activation process and improve clarity.

LiamSarsfield committed:
Merge branch 'add/boost/cache-preloading-on-cache-activation' into add/boost/cache-preloading-force-rebuild

LiamSarsfield committed:
Refactor Cache Preload Methods in Cache_Preload Class

- Renamed methods for clarity: `schedule_cornerstone_rebuild` to `schedule_cornerstone`, `rebuild` to `preload`, and `schedule_rebuild` to `schedule`.
- Updated action hooks to reflect new method names and improve consistency.
- Added logging for preload operations to enhance debugging capabilities.
- Simplified the activation process by ensuring the correct scheduling of cornerstones during activation.

LiamSarsfield committed:
Merge remote-tracking branch 'origin/trunk' into add/boost/cache-preloading-force-rebuild

LiamSarsfield committed:
Merge branch 'trunk' into add/boost/cache-preloading-on-cache-activation

LiamSarsfield committed:
Refactor Cache Preload Tests and Update Method Names

- Updated method names in tests for clarity: `preload_pages` to `request_pages`, `schedule_preload_cronjob` to `schedule`, and adjusted related assertions.
- Modified action hook references from `jetpack_boost_preload_pages` to `jetpack_boost_preload` for consistency.
- Added new tests for the `activate` and `schedule_cornerstone` methods to enhance coverage and ensure correct scheduling behavior.
- Improved comments for better understanding of test intentions.

LiamSarsfield committed:
Merge branch 'trunk' into add/boost/cache-preloading-on-cache-activation

LiamSarsfield committed:
Merge branch 'add/boost/cache-preloading-on-cache-activation' into add/boost/cache-preloading-force-rebuild

LiamSarsfield committed:
Add Page Cache Preloading changelog

LiamSarsfield committed:
Merge branch 'add/boost/cache-preloading-force-rebuild' of github.com:Automattic/jetpack into add/boost/cache-preloading-force-rebuild

LiamSarsfield committed:
Add custom cron intervals for Boost and update Cache_Preload class

- Introduced a new method `custom_cron_intervals` to add a 'twicehourly' schedule for cron jobs.
- Updated the `Cache_Preload` class to implement the `Has_Deactivate` contract and modified the `activate` and `deactivate` methods to use the new cron schedule.
- Removed the deprecated `schedule_cornerstone_cronjob` method and its associated tests to streamline functionality.

LiamSarsfield committed:
Refactor Jetpack Boost cron handling and Cache Preload class

- Splitting PR into smaller chuncks

LiamSarsfield committed:
Revert "Refactor Jetpack Boost cron handling and Cache Preload class"

This reverts commit 1f62fb019ea3a6ae4cbdf91d13a9ec19f4cb3998.

LiamSarsfield committed:
Update Cache_Preload_Test to use 'twicehourly' schedule for cron events

- Changed the expected function call from 'wp_schedule_single_event' to 'wp_schedule_event'.
- Updated the parameters to reflect the new cron schedule and maintain consistency with recent refactoring.

LiamSarsfield committed:
Merge remote-tracking branch 'origin/trunk' into add/boost/cache-preloading-force-rebuild

LiamSarsfield committed:
Add unit test for Cache_Preload class to verify cache rebuilding for specified URLs

- Implemented test_preload method to ensure cache is rebuilt correctly for given URLs.
- Mocked WordPress functions to simulate HTTP requests and cache invalidation actions.
- Enhanced test coverage for cache preloading functionality in the Cache_Preload class.

LiamSarsfield committed:
Merge branch 'add/boost/cache-preloading-force-rebuild' into add/boost/cache-preloading-scheduling

LiamSarsfield committed:
Enhance Cache Preload functionality by implementing cronjob scheduling for cornerstone pages

- Added `schedule_cornerstone_cronjob` method in `Cache_Preload` class to manage cronjob scheduling for preloading.
- Updated `activate` method to call the new scheduling method upon activation.
- Modified `Jetpack_Boost` class to trigger cache preloading when the page cache is enabled.
- Added unit tests to verify the scheduling behavior of the cronjob for cornerstone pages.

LiamSarsfield committed:
Merge remote-tracking branch 'origin/trunk' into add/boost/cache-preloading-scheduling

LiamSarsfield committed:
Merge branch 'trunk' into add/boost/cache-preloading-force-rebuild

LiamSarsfield committed:
Merge branch 'add/boost/cache-preloading-force-rebuild' into add/boost/cache-preloading-scheduling

LiamSarsfield committed:
Refactor Cache Preload scheduling for Cornerstone Pages

- Added `preload_cornerstone` method to handle cache rebuilding for Cornerstone Pages.
- Updated `schedule_cornerstone_cronjob` to schedule the new preload event.
- Removed deprecated `schedule_cornerstone_preload` method and adjusted related tests accordingly.
- Enhanced cache invalidation handling to trigger the new scheduling method.

LiamSarsfield committed:
Merge branch 'add/boost/cache-preloading-scheduling' of github.com:Automattic/jetpack into add/boost/cache-preloading-scheduling

LiamSarsfield commented on projects/plugins/boost/app/class-jetpack-boost.php:
```diff
@@ -179,6 +183,31 @@ public function handle_version_change() {
+	 *
+	 * @since $$next-version$$
+	 */
+	public function custom_cron_intervals( $schedules ) {
```
I'm unsure if this is the most appropriate place for creating this interval.

haqadn replied:
I think it's fine

LiamSarsfield committed:
Refactor Cache_Preload_Test to streamline cornerstone page handling

- Updated tests to use 'jetpack_boost_preload_cornerstone' for scheduling events.
- Removed unnecessary mocking of Cornerstone_Utils in multiple test cases.
- Simplified assertions related to cache invalidation and activation methods.

LiamSarsfield committed:
Add uninstall method to Cache_Preload for cleanup on plugin uninstallation

- Implemented `uninstall` method in `Cache_Preload` class to unschedule the cronjob for preloading Cornerstone Pages.
- Updated `Page_Cache_Setup` to call the new `uninstall` method during the plugin uninstallation process.

LiamSarsfield committed:
Refactor Cache_Preload methods for better cronjob management

- Updated `deactivate` method to use `wp_unschedule_hook` for unscheduling the preload event.
- Modified `uninstall` method to call `deactivate`, ensuring proper cleanup during plugin uninstallation.

LiamSarsfield committed:
Update uninstall method in Page_Cache_Setup to ensure Cache_Preload cleanup

- Moved Cache_Preload::uninstall() call to the beginning of the uninstall method for proper execution order.

LiamSarsfield committed:
Merge branch 'trunk' into add/boost/cache-preloading-force-rebuild

LiamSarsfield committed:
Merge branch 'add/boost/cache-preloading-force-rebuild' into add/boost/cache-preloading-scheduling

haqadn commented on projects/plugins/boost/app/modules/optimizations/page-cache/Page_Cache_Setup.php:
```diff
@@ -311,6 +311,7 @@ public static function deactivate() {
 	 */
 	public static function uninstall() {
 		self::deactivate();
+		Cache_Preload::uninstall();
```
Should this be in deactivate?

Ideally, the plugin activation/deactivation should call the activation/deactivation methods of all enabled modules. But, since it's not the case now, I thinking calling deactivate of cache preload here is fine. But, it should be on the deactivate method above.

LiamSarsfield replied:
I originally thought that `Page_Cache_Setup`'s `deactivate` would have triggered when Page Cache is disabled, but it's strictly only on plugin deactivation. I have noticed that deactivating Boost does not toggle Page Caching to off if Boost is activated again.
Because of this of this, what's your thoughts on calling Cache Preload's deactivation in `Page_Cache_Setup` `uninstall`? That's to prevent the Preload cronjob from being removed if someone deactivates Boost, and reactivates it with Page Caching enabled.

haqadn commented on projects/plugins/boost/app/modules/optimizations/page-cache/class-cache-preload.php:
```diff
@@ -59,20 +61,50 @@ public static function is_available() {
+	 * @since $$next-version$$
+	 * @return void
+	 */
+	public static function uninstall() {
+		self::deactivate();
+	}
```
Since `deactivate` is public, I think this method is redundant.

haqadn reviewed (changes requested):
I am suggesting a couple of cleanups. But, looks good!

LiamSarsfield committed:
Merge remote-tracking branch 'origin/trunk' into add/boost/cache-preloading-scheduling

LiamSarsfield committed:
Refactor Cache Preload Logic: Remove Uninstall Method and Ensure Proper Deactivation

- Removed the `uninstall` method from the `Cache_Preload` class to streamline the code.
- Updated the `Page_Cache_Setup` class to call `Cache_Preload::deactivate()` directly during cleanup, ensuring proper cache cleanup on plugin deactivation.
- Adjusted the `uninstall` method in `Page_Cache_Setup` to eliminate the call to the now-removed `Cache_Preload::uninstall()` method.

LiamSarsfield committed:
Refactor Page Cache Uninstall Logic: Move Cache Preload Deactivation

- Moved the call to `Cache_Preload::deactivate()` from the `setup` method to the `uninstall` method in `Page_Cache_Setup` to ensure proper cleanup during uninstallation.

LiamSarsfield committed:
Merge branch 'trunk' into add/boost/cache-preloading-scheduling

haqadn reviewed (approved):
LGTM
